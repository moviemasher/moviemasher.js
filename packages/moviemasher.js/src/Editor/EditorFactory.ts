import { Default } from "../Setup/Default"
import { EditorClass } from "./EditorClass"
import { Editor, EditorArgs, EditorOptions } from "./Editor"
export let editorSingleton: Editor
export const editorArgs = (options: EditorOptions = {}): EditorArgs => {
  return {
    autoplay: Default.masher.autoplay,
    precision: Default.masher.precision,
    loop: Default.masher.loop,
    fps: Default.masher.fps,
    volume: Default.masher.volume,
    buffer: Default.masher.buffer,
    ...options
  }
}
export const editorInstance = (options: EditorOptions = {}): Editor => {
  return editorSingleton = new EditorClass(editorArgs(options))
}
