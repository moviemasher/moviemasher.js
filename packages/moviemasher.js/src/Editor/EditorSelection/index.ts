export * from "./EditorSelection"
export * from "./EditorSelectionClass"
export * from "./EditorSelectionFactory"