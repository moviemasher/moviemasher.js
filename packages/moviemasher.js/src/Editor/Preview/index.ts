export * from "./AudioPreview"
export * from "./Preview"
export * from "./PreviewClass"
export * from "./TrackPreview"
