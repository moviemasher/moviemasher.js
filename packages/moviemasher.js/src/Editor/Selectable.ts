import { SelectType } from "../Setup/Enums"
import { Actions } from "./Actions/Actions"
import { SelectedItems } from "../Utility/SelectedProperty"

export interface Selectable { 
  selectables(): Selectables
  selectedItems(actions: Actions): SelectedItems 
  selectType: SelectType
}
export type Selectables = Selectable[]

export type SelectableRecord = {
  [index in SelectType]?: Selectable
}
export interface SelectTypesObject extends Record<string, SelectType[]> {}
