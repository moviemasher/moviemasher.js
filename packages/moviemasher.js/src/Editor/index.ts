export * from "./Actions"
export * from "./Editor"
export * from "./EditorClass"
export * from "./EditorFactory"
export * from "./Preview"
