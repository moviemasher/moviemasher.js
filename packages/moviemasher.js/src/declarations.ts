/* eslint-disable @typescript-eslint/no-namespace */
declare global { interface Window { webkitAudioContext: typeof AudioContext } }

export type Value = number | string
export type Scalar = boolean | Value | undefined
export type PopulatedString = string & { isEmpty: never }
export interface ValueObject extends Record<string, Value> {}
export interface NumberObject extends Record<string, number> {}
export interface BooleanObject extends Record<string, boolean> {}
export interface UnknownObject extends Record<string, unknown> {}
export interface StringObject extends Record<string, string> { }
export interface ScalarObject extends Record<string, Scalar> { }
export interface StringsObject extends Record<string, string[]> { }
export interface RegExpObject extends Record<string, RegExp> {}

export interface ObjectUnknown extends Record<string, UnknownObject> {}

export interface VisibleContextData extends ImageData {}
export interface VisibleContextElement extends HTMLCanvasElement { }

export interface AudibleContextData extends AudioContext {}
export interface Context2D extends CanvasRenderingContext2D {}

export interface Pixels extends Uint8ClampedArray {}
export interface LoadedImage extends HTMLImageElement {} // limited Image API in tests!
export interface LoadedVideo extends HTMLVideoElement {}
export interface LoadedAudio extends AudioBuffer {}
export interface LoadedFont extends FontFace { } // just { family: string } in tests!
export interface AudibleSource extends AudioBufferSourceNode {}

export type FfmpegSvgFilter = SVGFEFloodElement | SVGFEOffsetElement | SVGFEBlendElement | SVGClipPathElement
export type SvgFilter = FfmpegSvgFilter | SVGFEColorMatrixElement | SVGFEConvolveMatrixElement | SVGFEDisplacementMapElement
export type SvgFilters = SvgFilter[]
export type SvgItem = SVGElement 


export type SvgItems = SvgItem[]
export type SvgItemsTuple = [SvgItems, SvgItems]


export type SvgOrImage = SVGSVGElement | LoadedImage

export type VisibleSource = HTMLVideoElement | HTMLImageElement | SVGImageElement | HTMLCanvasElement

export type CanvasVisibleSource = VisibleSource | ImageBitmap | CanvasImageSource

export type Timeout = ReturnType<typeof setTimeout>
export type Interval = ReturnType<typeof setInterval>

export type LoadFontPromise = Promise<LoadedFont>
export type LoadImagePromise = Promise<LoadedImage>
export type LoadVideoPromise = Promise<LoadedVideo>
export type LoadAudioPromise = Promise<LoadedAudio>

export interface NumberConverter { (value: number): number }
export interface StringSetter { (value: string): void }
export interface NumberSetter { (value: number): void }
export interface BooleanSetter { (value: boolean): void }
export interface BooleanGetter { (): boolean }
export type EventHandler = (event: Event) => void 

export type AnyArray = any[]
export type JsonValue = Scalar | AnyArray | UnknownObject
export interface JsonObject extends Record<string, JsonValue | JsonValue[]> {}

export interface WithFrame {
  frame : number
}

export interface WithIndex {
  index : number
}

export interface WithTrack {
  trackNumber : number
}

export interface WithLabel {
  label : string
}

export interface Rgb {
  [index: string] : number
  r : number
  g : number
  b : number
}

export interface Rgba extends Rgb {
  a : number
}

export interface AlphaColor {
  color: string
  alpha: number
}

export interface AndType {
  type: string
}

export interface AndId {
  id: string
}

export interface AndLabel {
  label: string
}

export interface LabelAndId extends AndId, AndLabel {}

export interface WithError {
  error?: string
}

export interface AndTypeAndId extends AndType, AndId {}

export interface AndTypeAndValue extends AndType {
  value : number
}

export interface RgbObject {
  r: Value
  g: Value
  b: Value
}

export interface RgbaObject extends RgbObject {
  a: Value
}

export interface Rgb {
  r: number
  g: number
  b: number
}

export interface YuvObject {
  y: Value
  u: Value
  v: Value
}

export interface Yuv {
  y: number
  u: number
  v: number
}


// eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-explicit-any
export type Constrained<T = UnknownObject> = new (...args: any[]) => T

export interface GenericFactory<INSTANCE, INSTANCEOBJECT, DEFINITION, DEFINITIONOBJECT> {
  defaults?: DEFINITION[]
  definitionFromId(id : string) : DEFINITION
  definition(object: DEFINITIONOBJECT): DEFINITION
  instance(object : INSTANCEOBJECT) : INSTANCE
  fromId(id : string) : INSTANCE
}

export interface StartOptions {
  duration: number
  offset?: number
  start: number
}

// TODO: rename prefix to path and add query string parameters?
export interface Endpoint {
  protocol?: string
  prefix?: string
  host?: string
  port?: number
}

export interface UploadDescription {
  name: string
  type: string
  size: number
}

export interface InputParameter {
  key: string
  value: Value
}

export interface DescribedObject extends AndId, UnknownObject {
  icon?: string
  label?: string
}

export interface Described extends AndId {
  createdAt: string
  icon: string
  label: string
}



export const isCustomEvent = (value: any): value is CustomEvent => (
  value instanceof CustomEvent
)












