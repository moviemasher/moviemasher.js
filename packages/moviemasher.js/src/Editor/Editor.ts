import { StringRecord } from "../declarations"
import { IndexHandler } from "../Helpers/Select/Select"
import { PreviewItems } from "../Helpers/Svg/Svg"
import { Emitter } from "../Helpers/Emitter"
import { EditType, MasherAction } from "../Setup/Enums"
import { DataMashGetResponse } from "../Api/Data"
import { isMashAndMediaObject, MashMedia, Movable, MashMediaObject } from "../Media/Mash/Mash"
import { Effect } from "../Media/Effect/Effect"
import { Track } from "../Media/Mash/Track/Track"
import { Time, TimeRange } from "../Helpers/Time/Time"
import { Action } from "./Actions/Action/Action"
import { Clip, Clips } from "../Media/Mash/Track/Clip/Clip"
import { Actions } from "./Actions/Actions"
import { EditorSelection } from "./EditorSelection"
import { isObject } from "../Utility/Is"
import { errorThrow } from "../Helpers/Error/ErrorFunctions"
import { Rect } from "../Utility/Rect"
import { Size } from "../Utility/Size"
import { Media, MediaObject, MediaObjects, MediaArray } from "../Media/Media"
import { MediaCollection } from "../Base/MediaCollection"

export interface EditorIndex {
  layer?: number
  clip?: number
  track?: number
  effect?: number
}
export interface EditorArgs {
  autoplay: boolean
  buffer: number
  editType?: EditType
  baseUrl?: string
  mash?: MashMediaObject
  fps: number
  loop: boolean
  precision: number
  readOnly?: boolean
  dimensions?: Rect | Size | undefined
  volume: number
}

export interface EditorOptions extends Partial<EditorArgs> { }

export type ClipOrEffect = Clip | Effect

// export interface MashData extends Partial<DataMashGetResponse> { }

// export const isMashData = (value: any): value is MashData => (
//   isObject(value) && "mash" in value && isMashAndMediaObject(value.mash)
// )
// export function assertMashData(value: any, name?: string): asserts value is MashData {
//   if (!isMashData(value)) errorThrow(value, 'MashData', name)
// }

export interface Editor {
  actions: Actions
  addFiles(files: File[], editorIndex?: EditorIndex): Promise<Media[]>
  addClip(clip: Clip | Clips, editorIndex: EditorIndex): Promise<void>
  addMedia(object: MediaObject | MediaObjects, editorIndex?: EditorIndex): Promise<Media[]>
  addEffect: IndexHandler<Movable>
  addTrack(): void
  autoplay: boolean
  buffer: number
  can(action: MasherAction): boolean
  clips: Clips
  create(): Promise<void>
  currentTime: number
  // dataPutRequest(): Promise<DataPutRequest>
  dragging: boolean
  definitions: MediaArray
  definitionsUnsaved: MediaArray
  duration: number
  readonly mashMedia?: MashMedia
  editing: boolean
  editType: EditType
  eventTarget: Emitter
  fps: number
  goToTime(value: Time): Promise<void>
  handleAction(action: Action): void
  rect: Rect
  load(data: MashMediaObject): Promise<void>
  loop: boolean
  move(object: ClipOrEffect, editorIndex?: EditorIndex): void
  moveClip(clip: Clip, editorIndex?: EditorIndex): void
  moveEffect: IndexHandler<Movable>
  media: MediaCollection
  muted: boolean
  pause(): void
  paused: boolean
  play(): void
  position: number
  positionStep: number
  precision: number
  readOnly: boolean
  redo(): void
  redraw(): void
  removeClip(clip: Clip): void
  removeEffect: IndexHandler<Movable>
  // removeLayer(layer: Layer): void
  removeTrack(track: Track): void
  saved(temporaryIdLookup?: StringRecord): void
  readonly selection: EditorSelection
  svgElement: SVGSVGElement
  previewItems(enabled?: boolean): Promise<PreviewItems>
  time: Time
  timeRange: TimeRange
  undo(): void
  unload(): void
  updateDefinition(definitionObject: MediaObject, definition?: Media): Promise<void> 
  volume: number
}
