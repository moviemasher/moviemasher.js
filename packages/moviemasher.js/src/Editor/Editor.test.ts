import { UnknownObject } from "../declarations"
import { Default } from "../Setup/Default"
import { EditorClass } from "./EditorClass"
import { expectCanvasAtTime } from "../../../../dev/test/Utilities/expectMashSvg"
import { Emitter } from "../Helpers/Emitter"
import { editorInstance } from "./EditorFactory"
import { JestPreloader } from "../../../../dev/test/Utilities/JestPreloader"
import { Editor } from "./Editor"

import { timeFromSeconds } from "../Helpers/Time/TimeUtilities"
import { assertMash } from "../Edited/Mash/Mash"
import { EditType } from "../Setup/Enums"
import { ClipClass } from "../Edited/Mash/Track/Clip/ClipClass"

import visibleDefaultJson from "../Definitions/DefinitionObjects/content/default.json"

const createEditor = (): Editor => {
  return editorInstance({ editType: EditType.Mash, preloader: new JestPreloader()})
}
describe("EditorFactory", () => {

  describe("editorInstance", () => {
    test("returns EditorClass instance", () => {
      expect(createEditor()).toBeInstanceOf(EditorClass)
    })
  })
})

describe("Editor", () => {
  const values = Object.entries({
    loop: false,
    fps: 25,
    precision: 4,
    autoplay: true,
    volume: 0.4,
    buffer: 3,
  })
  const defaults = Object.entries(Default.masher)

  describe.each(defaults)("%s getter", (key, value) => {
    test("returns default", () => {
      const editor = createEditor() as unknown as UnknownObject
      if (key === 'fps') expect(editor[key]).toEqual(Default.masher.fps)
      else expect(editor[key]).toEqual(value)
    })
  })

  describe.each(values)("%s setter", (key, value) => {
    test("updates value", () => {
      const editor = createEditor() as unknown as UnknownObject
      expect(editor[key]).not.toEqual(value)
      editor[key] = value
      if (editor[key] !== value) throw new Error(`${key} ${value} ${editor[key]}`)
      expect(editor[key]).toEqual(value)
    })
  })

  describe("add", () => {
    test("returns promise that loads clip", async () => {
      const editor = createEditor()
      editor.load({ mash: {} })

      const clip = await editor.add(visibleDefaultJson)
      expect(clip).toBeInstanceOf(ClipClass)
      const { edited } = editor
      assertMash(edited)
      expectCanvasAtTime(editor)
    })
  })

  describe("eventTarget", () => {
    test("returns Emitter instance", () => {
      const editor = createEditor()
      const canvas = editor.eventTarget
      expect(canvas).toBeInstanceOf(Emitter)
    })
  })

  describe("goToTime", () => {
    test("returns what is set after resolving promise", async () => {
      const editor = createEditor()
      expect(editor.edited).toBeUndefined()
      editor.load({ mash: {} })
      const { edited } = editor
      assertMash(edited)

      await editor.add(visibleDefaultJson)
      expect(edited.frames).toEqual(30)
      const time = timeFromSeconds(2, editor.fps)
      await editor.goToTime(time)
      expect(editor.time).toEqual(time)
    })
  })


  describe("selectedItems", () => {
    test("returns expected properties", () => {
      const editor = createEditor()
      editor.load({ mash: {}, definitions: []})
      // masher.add(themeTextJson)
      const selectedItems = editor.selection.selectedItems()
      // console.log('selectedItems', selectedItems)
      expect(selectedItems.length).toEqual(2)
    })
  })

})
