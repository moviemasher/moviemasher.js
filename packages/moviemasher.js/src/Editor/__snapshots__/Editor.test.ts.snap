// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor add returns promise that loads clip: Unlabeled Mash 1`] = `"<svg width=\\"640\\" height=\\"480\\" viewBox=\\"0 0 640 480\\"><defs><marker id=\\"bounds-marker\\" markerUnits=\\"userSpaceOnUse\\" refX=\\"5\\" refY=\\"5\\" markerWidth=\\"10\\" markerHeight=\\"10\\" viewBox=\\"0 0 10 10\\"><rect width=\\"10\\" height=\\"10\\" fill=\\"currentColor\\" transform=\\"translate(5,5) scale(0.75,0.5625) translate(-5,-5)\\"></rect></marker><marker id=\\"shape-marker\\" markerUnits=\\"userSpaceOnUse\\" refX=\\"5\\" refY=\\"5\\" markerWidth=\\"10\\" markerHeight=\\"10\\" viewBox=\\"0 0 10 10\\"><ellipse cx=\\"5\\" cy=\\"5\\" rx=\\"5\\" ry=\\"5\\" fill=\\"currentColor\\" transform=\\"translate(5,5) scale(0.75,0.5625) translate(-5,-5)\\"></ellipse></marker></defs><svg width=\\"640\\" height=\\"480\\" viewBox=\\"0 0 640 480\\"><mask id=\\"mask-1\\"><g width=\\"480\\" height=\\"270\\" transform=\\"translate(0,0) scale(1.3333333333333333,1.7777777777777777)\\" transform-origin=\\"top left\\"><polygon points=\\"0,0 480,0 480,270, 0,270, 0,0\\" fill=\\"none\\" class=\\"bounds\\"></polygon><path d=\\"M 0.00 0.00 L 480.00 0.00 L 480.00 270.00 L 0.00 270.00 L 0.00 0.00\\" fill=\\"#FFFFFF\\" class=\\"shape\\"></path></g></mask><rect width=\\"640\\" height=\\"480\\" fill=\\"#FFFF00\\" x=\\"0\\" y=\\"0\\" mask=\\"url(#mask-1)\\" mask-mode=\\"alpha\\" filter=\\"url(#filter-2)\\"></rect><g width=\\"480\\" height=\\"270\\" transform=\\"translate(0,0) scale(1.3333333333333333,1.7777777777777777)\\" transform-origin=\\"top left\\"><polygon points=\\"0,0 480,0 480,270, 0,270, 0,0\\" fill=\\"none\\" class=\\"bounds selected\\" vector-effect=\\"non-scaling-stroke\\"></polygon><polygon points=\\"0,0 480,0 480,270, 0,270, 0,0\\" fill=\\"none\\" class=\\"bounds marker\\"></polygon><path d=\\"M 0.00 0.00 L 480.00 0.00 L 480.00 270.00 L 0.00 270.00 L 0.00 0.00\\" fill=\\"transparent\\" class=\\"shape marker\\"></path></g><filter id=\\"filter-2\\" filterUnits=\\"userSpaceOnUse\\"><feColorMatrix type=\\"matrix\\" values=\\"1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0\\"></feColorMatrix><feBlend mode=\\"normal\\"></feBlend></filter></svg></svg>"`;
